pkg/policy/policy.go

```diff
@@ -78,19 +77,45 @@ func (s *SearchContext) String() string {
 	for _, toLabel := range s.To {
 		to = append(to, toLabel.String())
 	}
+	// fmt.Sprintf() is well-known for not being opimal in terms
+	// of CPU and memory allocations. So we should avoide to use
+	// `fmt.Sprintf()` and replace it with `strings.Builder` or `bytes`.
+	// See https://github.com/cilium/cilium/issues/19571
+	var str strings.Builder
 	for _, dport := range s.DPorts {
 		if dport.Name != "" {
-			dports = append(dports, fmt.Sprintf("%s/%s", dport.Name, dport.Protocol))
+			str.Grow(len(dport.Name) + len(dport.Protocol) + 1)
+			str.WriteString(dport.Name)
+			str.WriteRune('/')
+			str.WriteString(dport.Protocol)
+			dports = append(dports, str.String())
 		} else {
-			dports = append(dports, fmt.Sprintf("%d/%s", dport.Port, dport.Protocol))
+			dport_Str := strconv.FormatUint(uint64(dport.Port), 10)
+			str.Grow(len(dport_Str) + len(dport.Protocol) + 1)
+			str.WriteString(dport_Str)
+			str.WriteRune('/')
+			str.WriteString(dport.Protocol)
+			dports = append(dports, str.String())
 		}
 	}
-	ret := fmt.Sprintf("From: [%s]", strings.Join(from, ", "))
-	ret += fmt.Sprintf(" => To: [%s]", strings.Join(to, ", "))
+	from_str := strings.Join(from, ", ")
+	to_str := strings.Join(to, ", ")
+	dport_str := strings.Join(dports, ", ")
+	var ret_str strings.Builder
+	// the length of 'From: [] => To: [] Ports: []' is 28
+	ret_str.Grow(len(from_str) + len(to_str) + len(dport_str) + 28)
+	ret_str.WriteString("From: [")
+	ret_str.WriteString(from_str)
+	ret_str.WriteRune(']')
+	ret_str.WriteString("=> To: [")
```

Comment:
This is missing a space. Would be easier to see if this was written as
```
    ret_str.WriteString("]=> To: [")
```
instead of two separate calls.